### ISSUE TITLE
Log::Any::Manager::_new_adapter_for_entry arguments are incorrect

### ISSUE THREAD

**Issue by karenetheridge** [elastic/elasticsearch-perl#61]
brought over from https://github.com/dagolden/Log-Any/issues/26

With Log::Any::Adapter 1.03 and Search::Elasticsearch 1.19:

the stacktrace is:

```
Odd number of elements in anonymous hash at /vagrant/extlib/lib/perl5/Log/Any/Adapter/Base.pm line 14.
        Log::Any::Adapter::Base::new("Log::Any::Adapter::Test", "/dev/null", "category", "elasticsearch.trace") called at /vagrant/extlib/lib/perl5/Log/Any/Manager.pm line 65
        Log::Any::Manager::_new_adapter_for_entry(Log::Any::Manager=HASH(0x1306a50), HASH(0x19bac6a0), "elasticsearch.trace") called at /vagrant/extlib/lib/perl5/Log/Any/Manager.pm line 31
        Log::Any::Manager::get_adapter(Log::Any::Manager=HASH(0x1306a50), "elasticsearch.trace") called at /vagrant/extlib/lib/perl5/Log/Any.pm line 81
        Log::Any::get_logger("Log::Any", "category", "elasticsearch.trace") called at /vagrant/extlib_more/lib/perl5/Search/Elasticsearch/Logger/LogAny.pm line 28
        Search::Elasticsearch::Logger::LogAny::_build_trace_handle(Search::Elasticsearch::Logger::LogAny=HASH(0x19b73c38)) called at (eval 13646)[/vagrant/extlib/lib/perl5/Sub/Quote.pm:3] line 18
        Search::Elasticsearch::Role::Logger::_assert_trace_handle(Search::Elasticsearch::Logger::LogAny=HASH(0x19b73c38)) called at (eval 13645)[/vagrant/extlib/lib/perl5/Sub/Quote.pm:3] line 17
        Search::Elasticsearch::Role::Logger::is_trace(Search::Elasticsearch::Logger::LogAny=HASH(0x19b73c38)) called at /vagrant/extlib_more/lib/perl5/Search/Elasticsearch/Role/Logger.pm line 54
        Search::Elasticsearch::Role::Logger::trace_request(Search::Elasticsearch::Logger::LogAny=HASH(0x19b73c38), Search::Elasticsearch::Cxn::HTTPTiny=HASH(0x649baa0), HASH(0x19d41680)) called at /vagrant/extlib_more/lib/perl5/Search/Elasticsearch/Transport.pm line 27
        Search::Elasticsearch::Transport::try {...} () called at /vagrant/extlib/lib/perl5/Try/Tiny.pm line 81
        eval {...} called at /vagrant/extlib/lib/perl5/Try/Tiny.pm line 72
        Try::Tiny::try(CODE(0x1b3361b0), Try::Tiny::Catch=REF(0x1b362748)) called at /vagrant/extlib_more/lib/perl5/Search/Elasticsearch/Transport.pm line 41
        Search::Elasticsearch::Transport::perform_request(Search::Elasticsearch::Transport=HASH(0x19d0e040), HASH(0x1b362850)) called at /vagrant/extlib_more/lib/perl5/Search/Elasticsearch/Role/Client.pm line 16
        Search::Elasticsearch::Role::Client::perform_request(Search::Elasticsearch::Client::Direct::Indices=HASH(0x19e61248), HASH(0x1b349250), "ignore", 404, "index", ARRAY(0x19f51cc0)) called at /vagrant/extlib_more/lib/perl5/Search/Elasticsearch/Role/Client/Direct.pm line 101
        Search::Elasticsearch::Role::Client::Direct::__ANON__[/vagrant/extlib_more/lib/perl5/Search/Elasticsearch/Role/Client/Direct.pm:102](Search::Elasticsearch::Client::Direct::Indices=HASH(0x19e61248), "ignore", 404, "index", ARRAY(0x19f51cc0)) called at /vagrant/extlib/lib/perl5/Elastic/Model/Role/Store.pm line 174
        Elastic::Model::Role::Store::get_aliases(Elastic::Model::__WRAPPED_15_::Elastic::Model::Store=HASH(0x1b330a88), "index", ARRAY(0x19f51cc0)) called at /vagrant/extlib/lib/perl5/Elastic/Model/Namespace.pm line 46
        Elastic::Model::Namespace::all_domains(Elastic::Model::__WRAPPED_1_::Elastic::Model::Namespace=HASH(0x19f57100)) called at /vagrant/extlib/lib/perl5/Elastic/Model/Role/Model.pm line 209
        Elastic::Model::Role::Model::_build_domain_namespace(CE::ElasticSearch::Model::Test=HASH(0x3d4b308)) called at native delegation method CE::ElasticSearch::Model::Test::_get_domain_namespace (get) of attribute _domain_namespace (defined at /vagrant/extlib/lib/perl5/Elastic/Model/Role/Model.pm line 135) line 13
        __ANON__[native delegation method CE::ElasticSearch::Model::Test::_get_domain_namespace (get) of attribute _domain_namespace (defined at /vagrant/extlib/lib/perl5/Elastic/Model/Role/Model.pm line 135):26](CE::ElasticSearch::Model::Test=HASH(0x3d4b308), "test") called at /vagrant/extlib/lib/perl5/Elastic/Model/Role/Model.pm line 225
        Elastic::Model::Role::Model::namespace_for_domain(CE::ElasticSearch::Model::Test=HASH(0x3d4b308), "test") called at /vagrant/extlib/lib/perl5/Elastic/Model/Role/Model.pm line 292
        Elastic::Model::Role::Model::domain(CE::ElasticSearch::Model::Test=HASH(0x3d4b308), "test") called at tlib/CE/Test.pm line 361
        CE::Test::create_esearch(CE::LeadRequestTransaction::State::Test=HASH(0x1975bc28)) called at tlib/CE/Test.pm line 218
```

The calls from my application (actually a test) are:

```
    my $model_class = 'CE::ElasticSearch::Model::Test';
    my $esearch_server    = CE::ElasticSearch::Object->new->instance;
    my $esearch_model     = $model_class->new(es => $esearch_server);
    my $esearch_namespace = $esearch_model->namespace('test');
    my $esearch_domain    = $esearch_model->domain('test');
```
..where CE::ElasticSearch::Object->new->instance boils down to:

```
Search::Elasticsearch->new(
          'trace_to' => [
                          'File',
                          '/dev/null'
                        ],
          'timeout' => 720,
          'nodes' => [
                       '127.0.0.1:9200'
                     ]
);
```

...except, when I run that code in a standalone script, it works fine...

I've data-dumped the full $esearch_model (it's quite large)... they look the same between the working and non-working code, but I'll re-verify that...

Please let me know what else I can provide that would be helpful!

**Comment by clintongormley** [comment]
Hi @karenetheridge 

So the weird thing is that your `trace_to` parameters claim to be using the `File` adapter, but the error comes when instantiating a `Test` adapter:

        Log::Any::Adapter::Base::new("Log::Any::Adapter::Test", "/dev/null", "category", "elasticsearch.trace") called at /vagrant/extlib/lib/perl5/Log/Any/Manager.pm line 65

Are you sure you're not overriding the Adapter somewhere in your tests?

**Comment by karenetheridge** [comment]
(sorry for not replying for so long.. I'm still investigating this issue, in between fighting fires.)

**Comment by karenetheridge** [comment]
Log::Any::Adapter::Test is getting loaded via another (unrelated) test class that does `use Log::Any::Test;`.  I'm tracing back int he code now to see what is inserting these adapter params (found by hooking Log::Any::Manager::set:

```
          'adapter_class' => 'Log::Any::Adapter::Test',
          'adapter_params' => [
                                '/dev/null'
                              ],
          'pattern' => qr/^elasticsearch\.trace$/
```

**Comment by karenetheridge** [comment]
I traced the call back to Search/Elasticsearch/Logger/LogAny.pm in _build_trace_handle, which calls:

`Log::Any::Adapter->set( { category => 'File'}, '/dev/null' );`

..will continue tracing farther back tomorrow.  I did find this entry in a config class for our app:

`$PARAMETERS{<some host>}{elastic_search}{trace_to} = ['File', '/dev/null'];`
...which is being picked up by our `MyApp::ElasticSearch::Model` class, but I'm not sure yet whether we're passing the config in the wrong format, or if it's being used incorrectly on the ES side. I shall return!

**Comment by clintongormley** [comment]
That is bizarre...  the `category` should be set by `trace_as`, not `trace_to, so I have no idea how `File` is ending up as the category...

Perhaps just dump `$self` in the `_build_trace_handle` method and paste it here?

**Comment by karenetheridge** [comment]
Here's the dump of $self in Search::Elasticsearch::Logger::LogAny::_build_trace_handle:

```
bless( {
                 'log_as' => 'elasticsearch.event',
                 'log_handle' => bless( {
                                          'adapter' => bless( {
                                                                'category' => 'elasticsearch.event'
                                                              }, 'Log::Any::Adapter::Test' ),
                                          'category' => 'elasticsearch.event',
                                          'formatter' => sub { "DUMMY" }
                                        }, 'Log::Any::Proxy::Test' ),
                 'serializer' => bless( {
                                          'JSON' => bless( do{\(my $o = '')}, 'Cpanel::JSON::XS' ),
                                          'mime_type' => 'application/json'
                                        }, 'Search::Elasticsearch::Serializer::JSON' ),
                 'trace_as' => 'elasticsearch.trace',
                 'trace_to' => [
                                 'File',
                                 '/dev/null'
                               ]
               }, 'Search::Elasticsearch::Logger::LogAny' );
```

...which results in Log::Any::Manager::set adding this to `$self->{entries}`:

```
        {
          'adapter_class' => 'Log::Any::Adapter::Test',
          'adapter_params' => [
                                '/dev/null'
                              ],
          'pattern' => qr/^elasticsearch\.trace$/
        };
```

From the documentation of Log::Any::Adapter::File, it looks like the argument list is correct (::File can take a single filename argument), but Log::Any::Adapter::Test doesn't do anything with the arg list, and so Log::Any:Adapter::Base::new gets it, and naively stuffs it into a hash.

So it's a bug with Log::Any::Adapter::Test then, I think?

(cc @dagolden)

**Comment by karenetheridge** [comment]
> So it's a bug with Log::Any::Adapter::Test then, I think?

This script reproduces the issue:

```
use strict;
use warnings;

{
    package Foo;
    use Log::Any::Adapter (
        'File', 'test.log', log_level => 'warn',
    );

    use Log::Any qw($log);
    BEGIN { $log->warn("test of file logger 2\n") }
}

{
    package Bar;
    use Log::Any::Test;    # should appear before 'use Log::Any'!
    use Log::Any::Adapter (
        'File', 'test.log', log_level => 'warn',
    );

    use Log::Any qw($log);

    BEGIN { $log->warn("test of test logger\n") }
}
```

I'm reopening https://github.com/dagolden/Log-Any/issues/26 :)

**Comment by karenetheridge** [comment]
Fixed by Log-Any-1.031-TRIAL!   thank you!
